### PR TITLE
Improved representation of AmiraMesh and HyperSurface files

### PR DESCRIPTION
**Pull request by hernot**
Hi
This pull request is not yet ready for merging. Never the less i started it for the following purpose

  - being able to run Tarvis checks as ar as configured, mainly related to Python2 support which might be broken by porting to Python3 which was weeks ago my initial starting point (see the long story below).

 - getting some feedback from you on my efforts to keep everything which might be necessary to keep your existing dependent codes operational without any changes while a the same time allowing me with the new structure to better understand the content of the pelora of different AmiraMesh and Hyper surface files i have to read and revive from an more than 10 Year old project.

 - discuss on how to structure the final pull-request/requests which than will make it into next release

 - discuss on whether to add, how and what tests  to the package ( have several scripts and plenty of different files i may pick some to be included for testing).

 - discuss on things i missed in the above line but are as equally important as the above

Here a very terse abstract of how all of this started and why i'm proposing things as are.
  
All of this started off some weeks ago when i was searching for a module for reading my huge collection of AmiraMesh and HyperSurface files created about 12 years ago. The only one which was able to read the data at all was your module after i fixed some things related to Python3 vs Python 2. This is why the module now supports Python3 and at least i tried to ensure Python2 support does not break. 
Looking at the data i noticed that the number returned for a binary encoded nodes Vector were quite ridiculously strange. Absolute values between 10^-10 to 10^38. Not quite plausible. Googling i found the Amira Reference guide (http://www1.udel.edu/ctcr/sites/udel.edu.ctcr/files/Amira%20Reference%20Guide.pdf) for Amira5 which describes all used file formats in detail. It stated that for both AmiraMesh and HyperSurface files, files which indicate a BINARY file format use bigendian numbers and files which have BINARY-LITTLE-ENDIAN format use littleendian numbers. By obeying to this convention suddenly my data was in plausible ranges, And i started to investigate and adopt codes to other not covered parts:
  `Field { float [3] FiledData } Linear(@1)`

Which basically tells that the data represented by the datapointer 1 is  a set of 3d Field vectors which have to
be linearly interpolated between the nodes of a tetrahedron mesh or surface mesh.
The only information about the mesh which this field belongs to is defined by the following two lines

` define Nodes 3000
 Nodes { float[3] NodeData} @1
`
So the Field line simply annotates the first datapointer but no information about the exempt that it has to consist of 3000 Nodes the vectors are rooted at.

In the current version the Header does not any more describe the header metadata as independent parts (designation, parameters, definitions, datapointers) but rather represents them in a tree  like structure. the designation values are now direct attributes of the header instance, the datapointers are now attributes of the array declarations (fromer definitions) and the parameters, wich  can now have infinite number of levels (some of my files have more than four).  Finally i changed the on demmand loading of the data streams such that no explicit call to read method of AmiraFile object is necessary any more. The data is implicitly loaded on the first access to the dedicated 'data' attribute of the corresponding metadata block or in case just the encoded raw data should be loaded without conversion into numpy arrays a call to stream_data is sufficient.
When decoded to numpy arrays than the resulting data is directly reshaped according to the dimension which is specified by array declaration and datatype specification for the data pointer.

for example 

`define Nodes 3000
Nodes {float[3] Coordinates} @1
`
when decoded will yield the folloing shape information for the underlying numpy array
`
$ <AmriaHeader>.Nodes.Coordinates.shape
(3000,3)
`
Lattices will be directly decoded into 3D numpy arrays 

Exempt for AmiraMeshDataStream and AmiraHxSurfaceDataStream classes which are now like the AmiraHeader class subclasses of the Block class any other specialized class for loading HyperSurface files i removed as they are not needed any more. The AmiraFile class is now a subclass of Amira Header providing read function and some other properties which i kept or readded for backward compatibility (generating a hopefully appropriate deprecation warning).

I changed the Block class such that all static class attributes are declared through the  `__slots__`

 special variable and any dynamic attribute added through add_attr will end up in the `__dict__` special variable which is included within the `__slots__` list (see https://docs.python.org/2.7/reference/datamodel.html?highlight=__slots__#__slots__ ) This at first allows to separate between dynamic and static attributes and it allows to use `__getattribute__` special method to trigger on demmand loading for data and stream_data attribures and all HyperSurface elements implicitly when called for the first time as the underlying private attributes are declared within the `__slots__` list but not yet created and thus cause an AttributeError exception which is captured by the `__getattr__` method and redirected to the StreamReader class which handles the rest.

### CODE REVIEW

**Comment by paulkorir** [comment]
Hi hernot,
Thanks for your work. I'm now beginning the processing of going through you changes. I'll be in touch constantly so that we can get a release out of this.
Regards,
Paul

File: ahds/ahds_common.py [hunk 1]
```diff
@@ -0,0 +1,791 @@
+# -*- coding: utf-8 -*-
```
Comment (paulkorir) [inline ahds/ahds_common.py:1]: While I understand the point of this module I'm wondering we can accomplish the same without too much plumbing. My main concern is debugging this code and testing it. If it is too complex then it will be a source of weakness.

File: ahds/data_stream.py [hunk 2]
```diff
@@ -55,124 +161,59 @@ def byterle_decoder(data, output_size):
...
-    :return numpy.array output: an array of ``numpy.uint8``
-    """
-    import zlib
-    data_stream = zlib.decompress(data)
-    output = numpy.array(struct.unpack('<{}B'.format(len(data_stream)), data_stream), dtype=numpy.uint8)
-    assert len(output) == data_size
-    return output
-
-
-def unpack_binary(data_pointer, definitions, data):
```
Comment (paulkorir) [inline ahds/data_stream.py:107]: Missing.

File: ahds/data_stream.py [hunk 3]
```diff
@@ -55,124 +161,59 @@ def byterle_decoder(data, output_size):
...
-        except AttributeError:
-            x, y, z = definitions.Lattice
-            data_length = x * y * z
-        
-    output = numpy.array(struct.unpack('<' + '{}'.format(data_type) * data_length, data)) # assume little-endian
-    output = output.reshape(data_length, data_dimension)
-    return output
-
-
-def unpack_ascii(data):
```
Comment (paulkorir) [inline ahds/data_stream.py:149]: Missing.

File: ahds/header.py [hunk 4]
```diff
@@ -8,27 +8,102 @@
...
+    >>> print(ah.Nodes)
+    print(ah.Nodes)
+     +-Nodes: (<ahds.ahds_common.Block object at 0x7f5dc1a74d08>)
+     |  |  +-Coordinates: (<ahds.data_stream.AmiraMeshDataStream object at 0x7f5dc1a97458>)
+     |  |  |  +-block: 1
+     |  |  |  +-array: Nodes(<ahds.ahds_common.Block object at 0x7f5dc1a74d08>)
+     |  |  |  +-type: float
+     |  |  |  +-dimension: 3
+     |  +-dimension: 3154
+    >>> print(ah.Tetrahedra.Nodes)
```
Comment (paulkorir) [inline ahds/header.py:65]: How do we know there is a Tetrahedra block here? It would be nice if one could get a snapshot of the whole structure then query individual blocks.

File: ahds/data_stream.py [hunk 5]
```diff
@@ -286,342 +327,877 @@ def __str__(self):
...
+                raise Exception("File encoding '{}' not supported".format(self._header.format))
+            return ft.partial(np.fromstring,dtype=_type_map[stream.type],sep="\n \t")
+        if self._header.format[0] not in ['B','b'] or self._header.format[1:6] not in ["INARY","inary"]:
+            raise Exception("File encoding '{}' not supported".format(self._header.format))
+        # check if data stream is compressed (HxByteRLE or HxZip)
+        _data_format = getattr(stream,"data_format",None)
+        # binary data encoded in bigendian (BINARY) or little endian (BINARY-LITTLER-ENDIAN)
+        # use according (True, False) section of _type_map lookuptable above
+        _endianess = self._header.format[6:] == "-LITTLE-ENDIAN"
+        if _data_format in ["HxByteRLE","hxbyterle"]:
```
Comment (paulkorir) [inline ahds/data_stream.py]: `if _data_format.lower() == "hxbyterle"`

File: ahds/header.py [hunk 6]
```diff
@@ -39,65 +114,165 @@
...
+block attibute in the above examples.
+
+The above approach can also be used to load Amira HyperSurface files.
+
+    >>> ah = AmiraHeader.from_file('<somfile>.surf')
+
+The restulting structure and behavior is the same as for AmiraMesh files.
+
+    >>> print(ah)
+    AMIRA HEADER (<somfile>.surf)
```
Comment (paulkorir) [inline ahds/header.py:127]: Very good idea to show nested blocks but some standardisation would make it easy for the user to understand.

File: ahds/grammar.py [hunk 7]
```diff
@@ -213,6 +82,91 @@ def number_seq(self, (tag, left, right, taglist), buffer_):
...
+
+    # in python3 and later open(<filename>,'rb') creates a binary file stream which has to be 
+    # explicitly decoded to unicode strings which are standard in python3 and later. Therefore
+    # any regular expression and string used to manipulate the raw stream data also has to be
+    # defined as byte string instead of regular raw pyhton string
+    _strip_lineend = b'\n'
+    _stream_delimiters = [
+        re.compile(b"(?:^|\n)@(?P<stream>\d+)\n",flags=re.S),
+        re.compile(r"(?:^|\n)\s*(?P<stream>(?:{}))(?:\s+(?:(?P<count>\d+)|(?P<name>\w+)))?(?:\s*\n|\s+)".format(_hyper_surface_entities).encode('ASCII')),
+        re.compile(b"^\s*}',re.I") # NOTE this is appried to reverese slice of stream_data therefore ^
```
Comment (paulkorir) [inline ahds/grammar.py]: Should `re.I` be inside the quotes?

File: ahds/ahds_common.py (hunk 1 above)
Comment (hernot) [reply ahds/ahds_common.py:1]: I do agree, that is why i wanted you to review first, so that we can identify together where things can be improved and with less achieved the same.
I already have from other projects some ideas which make things simplier

File: ahds/header.py (hunk 4 above)
Comment (hernot) [reply ahds/header.py:65]: For AmiraMesh file that is implicitly done as loading in a first step just reads the header without loading any data as done before. And not before accessing ah.Tetrahedra.Nodes.Data any data is loaded. There thechnically nothing changed. Exempt that Tetrahedra is the definition of array in the top lines of Amira mehs file while Nodes is the name of data form the data descriptor line

Tetrahedra { int [4] Nodes } @1

And as AmiraMesh has changed some time the array definition (need to check pdf) i decided to allways use the new name.
I add some test files, Any suggestions on how to structure test data tree and how to name it?

Reading encoded data from file and decoding are two distinct steps. The first is happening when anyway bypassing while looking for data block of interest. Deconding happens when data attribute is accessed for the first time.

File: ahds/header.py (hunk 6 above)
Comment (hernot) [reply ahds/header.py:127]: I'll try to come up with a proper suggestion in the next commit

File: ahds/grammar.py (hunk 7 above)
Comment (hernot) [reply ahds/grammar.py]: Oh shit. I has to be outside, is the case insensitive flag. Fix it in the next commit

File: ahds/data_stream.py (hunk 2 above)
Comment (hernot) [reply ahds/data_stream.py:107]: replaced it by numopy from buffer as it does the same with the advantage that data is directly available as numpy vector with appropriate data type including endianess and shape.

see line 207 in my code 

`return np.frombuffer(zlib.decompress(data),dtype=_np_ubytelitle,count=count)`

File: ahds/data_stream.py (hunk 3 above)
Comment (hernot) [reply ahds/data_stream.py:149]: numpy which is used in the end anyway offers fromstring method which does exactly this and has the advantage that data is directly in numpy data type and can be directly assigned the final shape of the resulting matrix. See select_decoder method below.

File: ahds/data_stream.py (hunk 5 above)
Comment (hernot) [reply ahds/data_stream.py]: Fix it in next commit.

File: ahds/__init__.py [hunk 8]
```diff
@@ -1,30 +1,47 @@
...
+    the AmiraHeader class will be merged into AmiraFile class and will be marked deprecated.
     """
+
+    __slots__ = ("_data_streams",)
     def __init__(self, fn, *args, **kwargs):
-        self._fn = fn
-        self._header = header.AmiraHeader.from_file(self._fn, *args, **kwargs)
-        self._data_streams = None # only populate on call to read() method
-        
+        super(AmiraFile,self).__init__(fn,*args,**kwargs)
```
Comment (paulkorir) [inline ahds/__init__.py:31]: I think there will be a signature mismatch here between this class and its parent.

File: ahds/header.py [hunk 9]
```diff
@@ -8,27 +8,102 @@
...
+        
 :: 
 
-    >>> print ah.data_pointers.attrs
-    ['data_pointer_1', 'data_pointer_2', 'data_pointer_3', 'data_pointer_4', 'data_pointer_5', 'data_pointer_6']
+    >>> print(ah.Nodes)
+    print(ah.Nodes)
+     +-Nodes: (<ahds.ahds_common.Block object at 0x7f5dc1a74d08>)
+     |  |  +-Coordinates: (<ahds.data_stream.AmiraMeshDataStream object at 0x7f5dc1a97458>)
+     |  |  |  +-block: 1
```
Comment (paulkorir) [inline ahds/header.py:60]: Help me understand: is there are scenario where we can have more than one value to the `block` field? Is this redundant?

File: ahds/header.py (hunk 4 above)
Comment (paulkorir) [reply ahds/header.py:65]: Perhaps we can have a data folder inside the ahds folder. At least for a start. Please choose then we can refine. Thanks very much.

**Comment by hernot** [comment]
Ok updated pullrequest. The latest commit greatly reduces complexity related to mix of properties and attributes. No MixedProperty any more no decprecation handlers thereof any more, no `@property` decorators any more no  `__getattr__`  anymore. All of them is now handled by `__getattribute__` methods most of which will vanish as soon as the attributes and former properties marked as deprecated are not needed any more for compoatibility with existing code. 
The `__getattribute__` approach has one big advantage it greatly simplifies load on access scheme especially in case of debuging as Attribute errors thrown by code which is trying to load the missing attribute keep their correct stack trace which would not be the case when using `__getattr__` and mixing that with `@property` getter methods.  Read only properties are simulated by `__setattr__` method limitting assignment to contexts of Block class and subclass methods unless `__setattr__` overloads `Block.__setattr__` method and calls add_attr to update attribute for now.

In total a lot less lines especially in adhs_common.py module. 

Further a tiny cosmetic as suggested by you now by explicitly indicating to add_attr method that attribute is back reference to parent Block this reference will be printed in python `__repr__` style in any case. Making it easy to directly identify parent references.

Fixed the little quoting error in gramma.py (saw just that i will need a second commit for fixing it on python2 branch too)

looking forward to your second review. Hope we can talk than also about adding test and possibly activating tarvis for automatic checking. Can provide some test data (surf, tetrahedra mesh, spreadsheets, marker sets, and other mostly binary)

File: ahds/header.py (hunk 4 above)
Comment (hernot) [reply ahds/header.py:65]: Ok i will try to do forth comming weekend when work on grant proposal is done had just a one day break for simplifying code. Will put some example data three and i also will add some test.py files inside a simple-tests folder we can remove afterwards i just do not want to collide with default github structure and at the same time i do not want to litter root

File: ahds/__init__.py (hunk 8 above)
Comment (hernot) [reply ahds/__init__.py:31]: lets check when proper tests are in place. In most cases only file name is needed.

File: ahds/header.py (hunk 9 above)
Comment (hernot) [reply ahds/header.py:60]: yes on HyperSurface files Patches section can have multiple block which represent a single surface patch
The same for optional Surfaces Section where each block joins one or more patches into one surface. By this a single HyperSurface file could contain multiple surfaces consisting of multiple patches each.

For Coordinates there will only be one block for which code currently does not make a distinction as this would have to be handled explicitly. It can but would make things more complicated.

And for AmiraMesh files this is simply the id of the data stream  iin the shown case in the file the Coordinates would be found under data stream @1 (data_definition["data_index"] or if not present data_definition["data_name"]). And yes here it is needed at least as long as it is required to provide data_pointers attribute/property. And i would keep it for debugging purpose as this would allow to directly lookup corresponding descriptor and data stream.

If you think that that should better be renamed than just let me know what better understandable name to choose.

File: ahds/ahds_common.py [hunk 10]
```diff
@@ -400,8 +197,28 @@ def move_attr(self,to,name):
         setattr(self,to,_tomove)
         delattr(self,name)
 
+    def __setattr__(self,name,value):
```
Comment (paulkorir) [inline ahds/ahds_common.py:167]: It's very hard to follow what's going on here. Is there another way to do this? I would suggest avoiding exotic interpreter objects especially because this is the core class about which everything is built.

Comment (hernot) [reply ahds/ahds_common.py:167]: I'd which there would be. What i have found is:

If sticking to attributes only which reduces headache of mixing properties with load on demmand semantics of attributes (dynamic properties are even bigger headache as properties are always to be defined on class and not instance) i haven't yet found any simpler and less magic approach to prevent attribute assignment through setattr and assignment operator = from outside class and subclass scope.

The only alternative would be a hidden `_lock` attribute which causes `__setattr__` to throw `AttributeError` whenever `_lock` is set to `True` and only accept assignment if set to `False`. But that would require to explicitly set `_lock` to `False` in in `add_attr` and any other methods of Block class and subclasses which wants to directly assining value to any attribute  directly including all`__init__` methods along the class hierarchie. And properly reset to previous state found on entering. Which is more error prone than having `Block.__setattr__` check the scope from within which the assignment is attempted and grant it if it comming from within Block class or subclass method.

Another approach i have neither tested nor tried and have no idea whether this is worth the efforts is by defining via custom meta class for Block. I wouldn't suggest this route for now as this adds complexity of defining Block meta class to code. Which i currently do not see any much benefit for the efforts needed adding even more magic than by simply inspecting callstack frame.

The only thing to reduce magic a bit could be to replace a variable by _calling_context_stack_frame and add comments which are anyway missing throughout more places of the code. Possibly also use methods from official python `inspect` module instead of calling hidden `sys._getframe` method but would not change anything in the logic.

Comment (hernot) [reply ahds/ahds_common.py:167]: I think i have found a good compromise. Any dynamic attributes defined by AmriaMesh and HyperSurface files which are stored inside `__dict__` can only be added or modified though a call to add_addr and renamed by calling `move_attr`. Any other attribute has to be declared thorugh `__slots__` and may be modified by direct assignment and through setattr calls.
Some protection is provided by pythons protected and private attribute mimickry (prepending by one ore two  `_`). They are not really read only but by convention nobody will modify them from outside class without very good reason. In addintion attributes starting with two `_` are rewritten to `_<ClassName><attributename>` Which is quite annoying when trying to display their content in debugsessions or apply filter operations for example in `__getattribute__` to them. Which is why i mean while stick to protected attributes and method names only where possible.

Any public attribute has to be explicitly write protected by explicitly overloading `__setattr__` method and comparing attribute name against list of read only attribute names and if found throwing an Attribute error while other wise passing on request.

**Comment by hernot** [comment]
See my second comment above concerning write protection.

Further i'd suggest to implement the `__contains__` method on the Block class hierarchy in order to allow the following test
```python
if Vertices in amirafileheader:
       print('hallo')
``` 
It would just report attributes defined within the file which are stored in class `__dict__`

**Comment by hernot** [comment]
Have tested the above suggestion. In general  works. Only during debugging in the rare case where an uncaught AttributeError occurs while trying to load HyperSurface files it may erroneously report as attribute defined by file a snot available. This is not a big issue as a break point can be placed on the line returning False in any case during debugging session.

Another thing i Just noticed. In AmiraMesh files defining Landmark sets (which can have multiple sets defined) Parameter lists defining parameters for individual landmark sets, are linked to their individual sets by  "Set-Id" parameter stating number of corresponding set. Actual sets are identified by

`Landmarks { float[3] Coordinates }` for first set and
`Landmarks { float[3] Coordiantes<n> }`  for any further set with 1 < n < NumSets

Same scheme applies to any other type of data defined by custom point sets derived from LandarkSets.

The issue now is that currently `Block` attributes containing special characters like '-' in their names for example "Set-Id" can only be accessed using 
```python
getattr(<containingblockinstance>, "Set-Id")
```

While any other attributes not containing these characters can be assessed like any other class attribute using '.' notation.

The question is now how to handle these parameter names allowed by AmiraMesh and HyperSurface file formats while the do not resemble a valid python class attribute identifier.  

1) replace all invalid characters by '_'. May cause name clash for parameters and attributes on carelessly defined or over sophisticatedly designed custom AmiraMesh objects and data types.
2) put a note in documentation that parameters, data arrays and data streams the names of which contain characters like '-', for exmaple "Set-Id" in HxLandarks files, can only be retrieved using `gettattr` based approach shown above.
3) other not so obvious (non) pythonish solution

File: ahds/ahds_common.py (hunk 10 above)
Comment (paulkorir) [reply ahds/ahds_common.py:167]: I like what I see. It's far more readable. Let me go through in detail then I'll get back to you.

Comment (hernot) [reply ahds/ahds_common.py:167]: I just noticed, the new solution has a slight bug between line 202 and 211. It allows setting of attribute even if not in slots of `self.__class__` which is the first along `__mro__` to be checked.  The whole block should read:

```python
for _ in (
    True
    for _slots in (
        getattr(_parent,"__slots__",None)
        for _parent in super(Block,self).__getattribute__("__class__").__mro__
    )
    if _slots is not None and name in _slots
):
    super(Block,self).__setattr__(name,value)
    return
```

This will be fixed in the next commit to assure all `__slots__` structures along `__mro__` of `self.__class__` are checked whether they contain the attribute or not. The first value returned by iterator in for loop is not used any further thus ingored using `_`.

**Comment by hernot** [comment]
Concerning tests i had a look at pytest. If deciding to use pytest and integrate with setuptools than i'd suggest to first finish this pullrequest and  than lateron when i have time i might open a new feature branch on my fork and a corresponding pull request just dealing with tests. Is cleaner. By the way how to send you the amira5 manual describing the file formats and structures or do you have it anyway?

**Comment by paulkorir** [comment]
> Concerning tests i had a look at pytest. If deciding to use pytest and integrate with setuptools than i'd suggest to first finish this pullrequest and than lateron when i have time i might open a new feature branch on my fork and a corresponding pull request just dealing with tests. Is cleaner. By the way how to send you the amira5 manual describing the file formats and structures or do you have it anyway?

I've had a look at pytest and it does look much easier to use than unittest. In fact, it can read unittest tests. I'll find time early next week to merge the pull request.

I have a copy of the Amira manual I downloaded from the Internet but it was for v5 or some old version. If there is a more recent one then I'd be happy to get it. Let me see if I can send you a private message with my email.

**Comment by hernot** [comment]
The one i sent you is the reference guide for Amira 5. I think there several kinds of manuals. The basic Manual, not much in there. The reference guide and than afaik there is also the developers manual comming along with the Amira developer package allowing to extend Amira with custom modlues (costs extra).

Ok I will commit my final changes and fixes. And ensure that documentation and comments is complete by toonight. or Latest tomorrow evening.

**Comment by hernot** [comment]
The one i sent you is the reference guide for Amira 5. I think there several kinds of manuals. The basic Manual, not much in there. The reference guide and than afaik there is also the developers manual comming along with the Amira developer package allowing to extend Amira with custom modlues (costs extra).

Ok made my final commit. Please let me know if i should stash some commits and cleanup commit messages.

**Comment by paulkorir** [comment]
Hi @hernot,
I've done a rebase and merge into the collab branch, which will eventually be merged into master after we've written tests and done cleanup.
Thank you very much for your hard work. It's greatly appreciated.
Paul

**Comment by hernot** [comment]
Thank you very much. I keep and rebase if it happens that i can help again.
